pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProvider.java

```diff
@@ -66,6 +69,10 @@ default String authenticate(AuthenticationDataSource authData) throws Authentica
         throw new AuthenticationException("Not supported");
     }
 
+    default List<String> authenticate(AuthenticationDataSource authData, boolean multiRoles) throws AuthenticationException {
```

Comment:
This is confusing to me `roles` are a set of permissions assigned to a principal. The job of authentication is to identify the user/app and get the principal/app id. For any given authData there should only be one identifier for it.

Roles should only be applicable with performing an authorization check. These roles can be accessed in the authorization provider by reading them from the authdata with will have the token.

I see other changes are references from the (https://github.com/apache/pulsar/pull/10375)
https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens. 

I think the `AuthenticationProviderToken.java` probably needs to be adjusted first.
The `sub` claim should be used to identify the principal or app id 
https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.2

Reply:
In some use cases, users need to use a set of roles for authorization instead of just one role. Authentication Provider should return all the roles and pass them to the authorization provider. In ServereCnx, we are using the auth Roles that returned from the authentication provider to authorize. I think it is necessary to let the authentication provider return a set of roles.

Reply:
I think the confusing part is we are using too many terms (role, principal id, app id, etc..) that are being overloaded. In my opinion the job of the authentication provider is to identity the client and this should be a single distinct value (principal id).  So I guess the question is what should the function be for the authorization provider? Should it identity the client or provide the client roles? I think it should identity the client because multiple clients could have the same roles and then it can become tricky from the logs and and auditing perspective what client is doing what specific action.

This change forces the broker server context and other classes to iterate over a list in multiple places and compare/sort lists. I think that complicates the code, makes it harder to maintain and reason about. The auth checks in the broker are already complicated and we should try to simply that. We should be trying to abstract and delegate these type of things to the auth providers, especially since this feature can already be done without any changes to the broker code or auth interfaces. In my opinion there are better ways to go about this.

Here are a few options:
1 - have the authorization provider discover the roles it should check for permissions. This requires none or minimal changes to the broker code. The authdata passed in to the authorization provider contains everything that is needed.

2 - Create an AuthInfo class or something similar that is returned by the authentication provider and that will contain a principal id and possible list of roles. However, this will involve changing the authorization interface to accept this new object and will be much larger change. The positive is this makes things much more flexible and extensible in the future.

On a side note there probably needs to be a cleanup of these terms (principal, role, app-id, client-id, etc...) within the code base so were consistent on what they mean.

Reply:
Thank you for your suggestions. This involves a lot of changes. Maybe we can implement this solution in another PR after this PR is merged?

Reply:
I think this should be done right and we should not push through something without thinking about the design and what role the authorization should play.  What you are trying to do can already be accomplished with the current framework (no code changes to any core pieces) by implementing a custom authorization provider. If we are going to make interfaces changes a design should be proposed to the community. I think it's better to think about this and come up with a design to support other potential use cases.

Reply:
I do agree that the nomenclature we are using is very confusing.  Some times we interchangeably use "principal" and "role" when by definition they can and should mean different things.  However, this seems like an existing issue not necessarily introduced in this PR?

Reply:
However, I don't understand why we need to pass in the boolean argument `multiRoles`.  There is already a method, i.e. `isSupportMultiRoles()`,  that returns whether the authenication provider supports multi-role or not